dex/encode/encode.go

```diff
@@ -0,0 +1,446 @@
+	return b
+}
+
+// Uint64Bytes converts the uint32 to a length-4, big-endian encoded byte slice.
+func Uint32Bytes(i uint32) []byte {
+	b := make([]byte, 4)
+	IntCoder.PutUint32(b, i)
+	return b
+}
+
+// Uint64Bytes converts the uint64 to a length-8, big-endian encoded byte slice.
+func Uint64Bytes(i uint64) []byte {
+	b := make([]byte, 8)
+	IntCoder.PutUint64(b, i)
+	return b
+}
+
+// CopySlice makes a copy of the slice.
+func CopySlice(b []byte) []byte {
+	newB := make([]byte, len(b))
+	copy(newB[:], b)
+	return newB
+}
+
+// RandomBytes returns a byte slice with the specified length of random bytes.
+func RandomBytes(len int) []byte {
+	bytes := make([]byte, len)
+	rand.Read(bytes)
+	return bytes
+}
+
+// ExtractPushes parses the linearly-encoded 2D byte slice into a slice of
+// slices.
+func ExtractPushes(b []byte) ([][]byte, error) {
+	pushes := make([][]byte, 0)
+	for {
+		if len(b) == 0 {
+			break
+		}
+		l := int(b[0])
+		b = b[1:]
+		if l == 255 {
+			if len(b) < 2 {
+				return nil, fmt.Errorf("2 bytes not available for uint16 data length")
+			}
+			l = int(IntCoder.Uint16(b[:2]))
+			b = b[2:]
+		}
+		if len(b) < l {
+			return nil, fmt.Errorf("data too short for pop of %d bytes", l)
+		}
+		pushes = append(pushes, b[:l])
+		b = b[l:]
+	}
+	return pushes, nil
+}
+
+// EncodePrefix encodes the order Prefix to a versioned blob.
+func EncodePrefix(p *order.Prefix) []byte {
+	return BuildyBytes{0}.
+		AddData(p.AccountID[:]).
+		AddData(Uint32Bytes(p.BaseAsset)).
+		AddData(Uint32Bytes(p.QuoteAsset)).
+		AddData([]byte{byte(p.OrderType)}).
+		AddData(Uint64Bytes(uint64(p.ClientTime.Unix()))).
+		AddData(Uint64Bytes(uint64(p.ServerTime.Unix())))
+}
+
+// DecodePrefix decodes the versioned blob to a *Prefix.
+func DecodePrefix(b []byte) (prefix *order.Prefix, err error) {
+	err = VersionedDecode("DecodePrefix", b, map[byte]ByteDecoder{
+		0: func(p [][]byte) error {
+			prefix, err = decodePrefix_v0(p)
+			return err
+		},
+	})
+	return prefix, err
+}
+
+// decodePrefix_v0 decodes the v0 payload into a *Prefix.
+func decodePrefix_v0(pushes [][]byte) (prefix *order.Prefix, err error) {
+	if len(pushes) != 6 {
+		return nil, fmt.Errorf("expected 6 prefix parts, got %d", len(pushes))
+	}
+	acctB, baseB, quoteB := pushes[0], pushes[1], pushes[2]
+	oTypeB, cTimeB, sTimeB := pushes[3], pushes[4], pushes[5]
+	if len(acctB) != account.HashSize {
+		return nil, fmt.Errorf("expected account ID length %d, got %d", account.HashSize, len(acctB))
+	}
+	var acctID account.AccountID
+	copy(acctID[:], acctB)
+	return &order.Prefix{
+		AccountID:  acctID,
+		BaseAsset:  IntCoder.Uint32(baseB),
+		QuoteAsset: IntCoder.Uint32(quoteB),
+		OrderType:  order.OrderType(oTypeB[0]),
+		ClientTime: time.Unix(int64(IntCoder.Uint64(cTimeB)), 0),
+		ServerTime: time.Unix(int64(IntCoder.Uint64(sTimeB)), 0),
+	}, nil
+}
+
+// EncodeTrade encodes the MarketOrder information, but not the Prefix
+// structure.
+func EncodeTrade(ord *order.MarketOrder) []byte {
```

Comment:
This concept of a "trade", which is the parts of the market order that aren't the prefix, keeps popping up for me. I implemented it in `msgjson` with the `Trade` as a separate type embedded by both market and limit order. I'd like to update the `order.Order` structures to reflect this model, and I think it has the potential to clean up the ever-growing `order.Order` interface in a few ways too.

Reply:
I'll follow up.